docs: update development setup

TomasTomecek opened the pull request:
Just let me know if the proposed changes are too drastic.

Also am not sure if the dockerfile should also be placed to a-r.

Signed-off-by: Tomas Tomecek <ttomecek@redhat.com>

csomh commented on docs/development-setup.md:
````diff
@@ -1,73 +1,137 @@
 
 ```
-$ docker-compose up -d
+$ docker build --no-cache --tag=buildroot --build-arg LATEST_CODE=true .
````
Shouldn't this have `REACTOR_SOURCE=git` instead of `LATEST_CODE=true`?

For those who are not masters of docker yet, would be nice to include who this step can be verified (`docker images` and look for *buildroot*) - error messages are not always obvious and can be easily missed.

TomasTomecek replied:
you're correct (and I thought I fixed all of those); I changed the variable name and semantics halfway through and apparently forgot to update all past references

csomh commented on docs/development-setup.md:
```diff
@@ -1,73 +1,137 @@
+
+In order to submit a build, you need to have a permission. If you started
+OpenShift with `oc cluster up`, there's a user `developer` set up with
+namespace `myproject`, this is what we'll be using.
```
I know this might be considered trivial, but the step to `oc login` mentioned explicitly here would be nice.

TomasTomecek replied:
To me, that's kind of redundant, but I certainly don't mind adding such info.

csomh commented on docs/development-setup.md:
````diff
@@ -1,73 +1,137 @@
-    builder_use_auth = false
-    use_auth=false
-    ```
+#### Registry
````
Shouldn't this be moved after the getting the token step? The registry IP should be used in `osbs.conf`.

TomasTomecek replied:
Good catch.

csomh commented on docs/development-setup.md:
````diff
@@ -1,73 +1,137 @@
+address of the registry by issuing following command:
+
+```
+$ oc describe service docker-registry
````
I could not make this work, and couldn't figure it out what is missing.
I've did `oc login -u developer`.
`oc get svc` gives me nothing and `oc status` gives me this:
```
In project My Project (myproject) on server https://192.168.122.37:8443

You have no services, deployment configs, or build configs.
Run 'oc new-app' to create an application.
```

TomasTomecek replied:
You're absolutely correct. I've just realized that this is wrong (I deployed another registry by accident).

TomasTomecek commented:
Thanks for review. Great suggestions!

lcarva commented on docs/development-setup.md:
````diff
@@ -1,73 +1,158 @@
 
 ```
 $ git clone git@github.com:projectatomic/osbs-client.git
 $ cd osbs-client
````
Maybe add `python setup.py develop` for functional parity with `dnf` approach below?

TomasTomecek replied:
Great suggestion! I actually have my own aliases for this problem. That's why I forgot about the CLI when just cloning the repo.

lcarva commented on docs/development-setup.md:
```diff
@@ -28,6 +28,9 @@ For local development, we advise to clone this git repo:
 $ cd osbs-client
+$ pip3 install -r ./requirements.txt
+$ pip3 install -r ./requirements-py3.txt
+$ python3 ./setup.py develop
```
Shouldn't this line take care of the requirements?

TomasTomecek commented:
Is any of the issues still unresolved? If not, can we have this merged?

TomasTomecek commented:
I think I resolved most of the issues. Only one remains: pushing images.

Build container needs to use `builder` service account for authentication. This can be done via mounting dockercfg secret in the build container (when present in home, no changes are needed in a-r).

I'm assuming this could be done by adding new configuration option to bind the secret in the pod. Preferably to `~/.docker/config.json`.

Honestly, I won't have time to implement that. If you have time to do that, great! If not, I remove the registry part from the docs.

twaugh commented:
Doesn't the `builder` service account get used by default for builds? I don't think we have to do anything special for that.

TomasTomecek commented:
This is about actual image push. In order to do that successfully, you need to authenticate with the registry. OpenShift is able then authorize your request, if you authenticate with correct credentials. Hence you need the credentials in build container (maybe they are mounted there by default, haven't checked this).

We would have to add auth bits [here](https://github.com/projectatomic/atomic-reactor/blob/13d0e3e867752c1715e4360a6aab6cd5d12dd965/atomic_reactor/core.py#L481).

KarstenHopp commented:
Two observations from reading development-setup.md:
'Obtaining token' should probably go to the 'config file' section with an explanation that the placeholder token in your example needs to be replaced with it.

The 'registry' section is in the wrong order imho. First there is the 'oc describe service docker-registry' command, which will fail as you're logged in as developer after installation.
So the correct order would be 
 oc login -u system:admin
 oc project default
 oc describe service docker-registry
and then (missing)
 oc login -u developer

lcarva commented on docs/development-setup.md:
````diff
@@ -1,73 +1,164 @@
+
+ 2. You can start a single-node cluster like this
+    ```
+    $ oc cluster up
````
In addition, the following is required to allow builder ServiceAccount to perform operations from within builder container:
`oc adm policy add-role-to-user admin system:serviceaccount:myproject:builder`

For instance, without this, the plugin `store_metadata_in_osv3` fails with 403 error:  `User "system:serviceaccount:myproject:builder" cannot get builds in project "myproject"`,

twaugh replied:
Does the 'edit' role work for this or is 'admin' required?

TomasTomecek replied:
Luiz, you're correct! And Tim as well, `edit` is sufficient.

TomasTomecek commented:
Hey guys, I would love to update the PR this week, so hopefully we can merge before Christmas or in early January.

TomasTomecek commented:
This should be ready for review now. I think I addressed all the comments. Unfortunately I had to "cut" the registry part b/c it just didn't work and would require some non-trivial code changes.

Thank you to everyone who helped with reviews: @KarstenHopp @lcarva @twaugh @csomh

TomasTomecek commented:
Luiz, thanks for your thorough review!

I agree with all your points, I bumped into 3) 4) (and probably 2) but haven't noticed it).

lcarva commented:
@TomasTomecek tried fixing conflict in README.md file, but github still complains. Could you rebase so we can merge this?
